plugins/modem-manager/fu-mm-device.c

```diff
@@ -1742,45 +1730,112 @@ fu_mm_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+					   &error_local);
 
 	/* firmware branch AT command may fail if not implemented,
 	 * clear error if not supported */
-	if (self->branch_at != NULL) {
-		g_autoptr(GError) error_branch = NULL;
-		if (!fu_mm_device_at_cmd(self, self->branch_at, TRUE, &error_branch))
-			g_debug("unable to get firmware branch: %s", error_branch->message);
+	if (locker == NULL)
+		g_debug("failed to open AT port: %s", error_local->message);
+	else if (!fu_mm_device_at_cmd(self, self->branch_at, TRUE, &error_local))
+		g_debug("unable to get firmware branch: %s", error_local->message);
+}
+
+static void
+fu_mm_device_setup_secboot_status_quectel(FuMmDevice *self)
+{
+	if (self->port_at != NULL) {
+		g_autoptr(FuDeviceLocker) locker = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		/* Create IO channel to send AT commands to the modem */
+		locker = fu_device_locker_new_full(self,
+						   (FuDeviceLockerFunc)fu_mm_device_io_open,
+						   (FuDeviceLockerFunc)fu_mm_device_io_close,
+						   &error_local);
+		if (locker == NULL)
+			g_debug("failed to open AT port: %s", error_local->message);
+		else if (!fu_mm_device_at_cmd(self, "AT+QSECBOOT=\"status\"", TRUE, &error_local))
+			g_debug("failed to send AT+QSECBOOT command: %s", error_local->message);
+	} else {
+		const gchar *version = fu_device_get_version(FU_DEVICE(self));
+		g_autofree gchar *name = NULL;
+
+		struct {
+			const gchar *name;
+			const gchar *version;
+		} secboot[] = {{"EM05GF", "EM05GFAR07A07M1G_01.005.01.005"},
+			       {"EM05CE", "EM05CEFCR08A16M1G_LNV"},
+			       {NULL, NULL}};
+
+		/* find model name and compare with table from Quectel */
+		if (version == NULL)
+			return;
+		name = g_strndup(version, 6);
+		for (guint i = 0; secboot[i].name != NULL; i++) {
+			if (g_strcmp0(name, secboot[i].name) == 0) {
+				if (fu_version_compare(version,
+						       secboot[i].version,
+						       FWUPD_VERSION_FORMAT_PLAIN) >= 0) {
+					fu_device_add_flag(FU_DEVICE(self),
+							   FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+				} else {
+					fu_device_add_flag(FU_DEVICE(self),
+							   FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+				}
+				return;
+			}
+		}
 	}
+}
 
-	/* success */
-	return TRUE;
+static void
+fu_mm_device_setup_secboot_status(FuDevice *device)
+{
+	GPtrArray *vendors;
+	FuMmDevice *self = FU_MM_DEVICE(device);
+
+	vendors = fu_device_get_vendor_ids(device);
+	if (vendors != NULL && vendors->len != 0) {
+		for (guint i = 0; i < vendors->len; i++) {
+			const gchar *vendor = g_ptr_array_index(vendors, i);
+			/* Quectel modem */
+			if (g_strcmp0(vendor, "USB:0x2C7C") == 0 ||
+			    g_strcmp0(vendor, "PCI:0x1EAC") == 0) {
+				fu_mm_device_setup_secboot_status_quectel(self);
+				break;
+			}
+		}
+	}
 }
 
 static gboolean
 fu_mm_device_setup(FuDevice *device, GError **error)
 {
 	FuMmDevice *self = FU_MM_DEVICE(device);
 
-	if (self->port_at != NULL) {
-		if (!fu_mm_device_setup_branch_at(self, error))
-			return FALSE;
-	}
+	fu_mm_device_setup_branch_at(self);
```

Comment:
@ivanmikh Since you said that you cannot disable secboot at all, branches should be blocked IMO since you can't switch at all. Or at least disable the `community` option if Quectel decides to use branches at some point.

Reply:
Or the community one should be able to sign their firmware?

Reply:
ack; if the device is signed then my definition it can't support branch-switching.

Reply:
okay, i added a check for signed firmware before running firmware branch AT command